File: examples/fortran/distributed_vector.f90
! Copyright (c) 2026 Bryce M. Westheimer
! SPDX-License-Identifier: BSD-3-Clause

!> @file distributed_vector.f90
!> @brief Distributed vector operations with DTL Fortran bindings
!>
!> Demonstrates:
!> - Creating distributed vectors
!> - Accessing local data
!> - Computing local statistics
!>
!> Compile with:
!>   gfortran -c dtl_fortran.f90
!>   gfortran -L../../build/src/bindings/c distributed_vector.f90 dtl_fortran.o -ldtl_c -o distributed_vector
!>
!> Run with:
!>   ./distributed_vector
!>   mpirun -np 4 ./distributed_vector

program distributed_vector_example
    use, intrinsic :: iso_c_binding
    use dtl_fortran
    implicit none

    type(c_ptr) :: ctx, vec
    integer(c_int32_t) :: status
    integer(c_int32_t) :: rank, size
    integer(c_int64_t) :: global_size, local_size, local_offset
    real(c_double), pointer :: data(:)
    type(c_ptr) :: data_ptr
    integer(c_int64_t) :: i
    real(c_double) :: local_sum, local_min, local_max, local_mean
    real(c_double), target :: fill_value

    ! Create context
    status = dtl_context_create_default(ctx)
    if (.not. is_success(status)) then
        print '(A)', 'ERROR: Failed to create context'
        stop 1
    end if

    rank = dtl_context_rank(ctx)
    size = dtl_context_size(ctx)

    print '(A,I0,A,I0,A)', 'Rank ', rank, ' of ', size, ': Distributed Vector Example'

    ! Create a distributed vector of 1000 doubles
    global_size = 1000
    status = dtl_vector_create(ctx, DTL_DTYPE_FLOAT64, global_size, vec)
    if (.not. is_success(status)) then
        print '(A)', 'ERROR: Failed to create vector'
        stop 1
    end if

    ! Query vector properties
    local_size = dtl_vector_local_size(vec)
    local_offset = dtl_vector_local_offset(vec)

    print '(A,I0,A,I0,A,I0,A,I0)', 'Rank ', rank, ': global_size=', &
        dtl_vector_global_size(vec), ', local_size=', local_size, &
        ', local_offset=', local_offset

    ! Get local data pointer and convert to Fortran pointer
    data_ptr = dtl_vector_local_data_mut(vec)
    call c_f_pointer(data_ptr, data, [local_size])

    ! Initialize: each element is its global index
    do i = 1, local_size
        data(i) = real(local_offset + i - 1, c_double)
    end do

    ! Compute local statistics
    local_sum = 0.0_c_double
    local_min = data(1)
    local_max = data(1)

    do i = 1, local_size
        local_sum = local_sum + data(i)
        if (data(i) < local_min) local_min = data(i)
        if (data(i) > local_max) local_max = data(i)
    end do

    local_mean = local_sum / real(local_size, c_double)

    print '(A,I0,A,F10.2,A,F10.2,A,F10.2,A,F10.2)', &
        'Rank ', rank, ': sum=', local_sum, ', min=', local_min, &
        ', max=', local_max, ', mean=', local_mean

    ! Synchronize
    status = dtl_vector_barrier(vec)
    if (.not. is_success(status)) then
        print '(A)', 'ERROR: Barrier failed'
        stop 1
    end if

    ! Fill with a constant value
    fill_value = 42.0_c_double
    status = dtl_vector_fill_local(vec, c_loc(fill_value))
    if (.not. is_success(status)) then
        print '(A)', 'ERROR: Fill failed'
        stop 1
    end if

    ! Verify fill
    if (all(abs(data - fill_value) < 1.0e-10_c_double)) then
        print '(A,I0,A)', 'Rank ', rank, ': fill verification PASSED'
    else
        print '(A,I0,A)', 'Rank ', rank, ': fill verification FAILED'
    end if

    ! Cleanup
    call dtl_vector_destroy(vec)
    call dtl_context_destroy(ctx)

    if (rank == 0) then
        print '(A)', ''
        print '(A)', 'Done!'
    end if

end program distributed_vector_example

File: examples/c/distributed_vector.c
// Copyright (c) 2026 Bryce M. Westheimer
// SPDX-License-Identifier: BSD-3-Clause

/**
 * @file distributed_vector.c
 * @brief Distributed vector operations with DTL C bindings
 *
 * Demonstrates:
 * - Creating distributed vectors
 * - Accessing local data
 * - Fill operations
 * - Computing local statistics
 *
 * Compile with:
 *   gcc -I../../include -L../../build/src/bindings/c distributed_vector.c -ldtl_c -lm -o distributed_vector
 *
 * Run with:
 *   ./distributed_vector
 *   mpirun -np 4 ./distributed_vector
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <dtl/bindings/c/dtl.h>

#define CHECK_STATUS(status, msg) \
    do { \
        if (!dtl_status_ok(status)) { \
            fprintf(stderr, "%s: %s\n", msg, dtl_status_message(status)); \
            exit(1); \
        } \
    } while(0)

int main(int argc, char** argv) {
    (void)argc;
    (void)argv;

    dtl_context_t ctx = NULL;
    dtl_vector_t vec = NULL;
    dtl_status status;

    /* Create context */
    status = dtl_context_create_default(&ctx);
    CHECK_STATUS(status, "Failed to create context");

    dtl_rank_t rank = dtl_context_rank(ctx);
    dtl_rank_t size = dtl_context_size(ctx);

    printf("Rank %d of %d: Distributed Vector Example\n", rank, size);

    /* Create a distributed vector of 1000 doubles */
    dtl_size_t global_size = 1000;
    status = dtl_vector_create(ctx, DTL_DTYPE_FLOAT64, global_size, &vec);
    CHECK_STATUS(status, "Failed to create vector");

    /* Query vector properties */
    printf("Rank %d: global_size=%lu, local_size=%lu, local_offset=%ld\n",
           rank,
           (unsigned long)dtl_vector_global_size(vec),
           (unsigned long)dtl_vector_local_size(vec),
           (long)dtl_vector_local_offset(vec));

    /* Get local data pointer */
    double* data = (double*)dtl_vector_local_data_mut(vec);
    dtl_size_t local_size = dtl_vector_local_size(vec);
    dtl_index_t offset = dtl_vector_local_offset(vec);

    /* Initialize: each element is its global index */
    for (dtl_size_t i = 0; i < local_size; i++) {
        data[i] = (double)(offset + (dtl_index_t)i);
    }

    /* Compute local statistics */
    double local_sum = 0.0;
    double local_min = data[0];
    double local_max = data[0];

    for (dtl_size_t i = 0; i < local_size; i++) {
        local_sum += data[i];
        if (data[i] < local_min) local_min = data[i];
        if (data[i] > local_max) local_max = data[i];
    }

    double local_mean = local_sum / (double)local_size;

    printf("Rank %d: sum=%.2f, min=%.2f, max=%.2f, mean=%.2f\n",
           rank, local_sum, local_min, local_max, local_mean);

    /* Synchronize */
    status = dtl_vector_barrier(vec);
    CHECK_STATUS(status, "Barrier failed");

    /* Fill with a constant value */
    double fill_value = 42.0;
    status = dtl_vector_fill_local(vec, &fill_value);
    CHECK_STATUS(status, "Fill failed");

    /* Verify fill */
    int all_correct = 1;
    for (dtl_size_t i = 0; i < local_size; i++) {
        if (fabs(data[i] - fill_value) > 1e-10) {
            all_correct = 0;
            break;
        }
    }
    printf("Rank %d: fill verification %s\n", rank, all_correct ? "PASSED" : "FAILED");

    /* Cleanup */
    dtl_vector_destroy(vec);
    dtl_context_destroy(ctx);

    if (rank == 0) {
        printf("\nDone!\n");
    }

    return 0;
}
